client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentEventReaderImpl.java

```diff
@@ -0,0 +1,380 @@
+    @GuardedBy("$lock")
+    private long endOffset;
+    @GuardedBy("$lock")
+    private final ReadState readState;
+    @GuardedBy("$lock")
+    private CompletableFuture<SegmentRead> outstandingRequest = null;
+
+    AsyncSegmentEventReaderImpl(AsyncSegmentInputStream asyncInput, long startOffset) {
+        this(asyncInput, startOffset, UNBOUNDED_END_OFFSET, DEFAULT_READ_LENGTH);
+    }
+
+    AsyncSegmentEventReaderImpl(AsyncSegmentInputStream asyncInput, long startOffset, long endOffset, int readLength) {
+        Preconditions.checkArgument(startOffset >= 0);
+        Preconditions.checkNotNull(asyncInput);
+        Preconditions.checkNotNull(endOffset, "endOffset");
+        Preconditions.checkArgument(endOffset > startOffset + WireCommands.TYPE_PLUS_LENGTH_SIZE,
+                "Invalid end offset.");
+        this.asyncInput = asyncInput;
+        this.cursorOffset = startOffset;
+        this.endOffset = endOffset;
+
+        /*
+         * The logic for determining the read length and buffer size are as follows.
+         * If we are reading a single event, then we set the read length to be the size
+         * of the event plus the header.
+         *
+         * If this input stream is going to read many events of different sizes, then
+         * we set the read length to be equal to the max write size and the buffer
+         * size to be twice that. We do it so that we can have at least two events
+         * buffered for next event reads.
+         */
+        this.readLength = Math.min(DEFAULT_READ_LENGTH, readLength);
+
+        this.readState = new ReadState();
+    }
+
+    @Override
+    public Segment getSegmentId() {
+        return asyncInput.getSegmentId();
+    }
+
+    @Override
+    @Synchronized
+    public void close() {
+        log.trace("Closing {}", this);
+        readState.cancel();
+        asyncInput.close();
+    }
+
+    @Override
+    public boolean isClosed() {
+        return asyncInput.isClosed();
+    }
+
+    // region Cursor
+
+    @Override
+    @Synchronized
+    public void setOffset(long offset) {
+        log.trace("SetOffset {}", offset);
+        Preconditions.checkArgument(offset >= 0);
+        Exceptions.checkNotClosed(asyncInput.isClosed(), this);
+        if (offset != this.cursorOffset) {
+            readState.cancel();
+            this.cursorOffset = offset;
+        }
+    }
+
+    @Override
+    @Synchronized
+    public long getOffset() {
+        return cursorOffset;
+    }
+
+    @Override
+    @Synchronized
+    public void setEndOffset(long offset) {
+        log.trace("SetEndOffset {}", offset);
+        Preconditions.checkArgument(offset >= 0);
+        Exceptions.checkNotClosed(asyncInput.isClosed(), this);
+        this.endOffset = offset;
+    }
+
+    @Override
+    @Synchronized
+    public long getEndOffset() {
+        return endOffset;
+    }
+
+    @Override
+    @Synchronized
+    public CompletableFuture<ByteBuffer> readAsync() throws EndOfSegmentException {
```

Comment:
What are the expected semantics when this is called twice in succession? (Without waiting on the futures?)

Reply:
In order for that not to have a race when updating the offset the callback would have to be completed under lock.

Reply:
If the caller calls `setOffset`, any outstanding future that was returned from `readAsync` is cancelled before updating the internal state.

Reply:
I don't mean a caller calling setOffset. I mean when the offset advances as a result of the read completing.